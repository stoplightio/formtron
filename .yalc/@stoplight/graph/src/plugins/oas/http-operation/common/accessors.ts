import unionBy = require('lodash/unionBy');

export function getOasParameters<ParamType extends { name: string; in: string }>(
  operationParameters: ParamType[] | undefined,
  pathParameters: ParamType[] | undefined
) {
  return unionBy(operationParameters, pathParameters, (parameter: ParamType) => `${parameter.name}-${parameter.in}`);
}
