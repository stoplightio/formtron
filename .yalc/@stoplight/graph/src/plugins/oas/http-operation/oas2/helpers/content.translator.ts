import { IHttpContent } from '@stoplight/types';
import { ISchema } from '@stoplight/types/schema';

export function translateToContent(schema: ISchema, mediaTypes: string[]): { [mediaType: string]: IHttpContent } {
  return mediaTypes.reduce((contentDict, mediaType) => {
    contentDict[mediaType] = {
      mediaType,
      schema,
    };
    return contentDict;
  }, {});
}
