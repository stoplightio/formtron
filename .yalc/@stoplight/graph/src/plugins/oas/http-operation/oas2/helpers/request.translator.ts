import { IHttpRequestParams } from '@stoplight/types';
import { Parameter } from 'swagger-schema-official';
import { isBodyParameter, isFormDataParameter, isHeaderParameter, isPathParameter, isQueryParameter } from './guards';
import {
  translateToBodyParameter,
  translateToFormDataParameter,
  translateToHeaderParam,
  translateToPathParameter,
  translateToQueryParameter,
} from './params.translator';

export function translateToRequest(parameters: Parameter[], consumes: string[]): IHttpRequestParams {
  return parameters.reduce((request: IHttpRequestParams, parameter: Parameter) => {
    if (isBodyParameter(parameter)) {
      request.body = translateToBodyParameter(parameter, consumes);
    } else if (isFormDataParameter(parameter)) {
      request.body = translateToFormDataParameter(parameter, request.body, consumes);
    } else if (isQueryParameter(parameter)) {
      const queryParameter = translateToQueryParameter(parameter, consumes);
      request.query = (request.query || []).concat(queryParameter);
    } else if (isPathParameter(parameter)) {
      const pathParameter = translateToPathParameter(parameter, consumes);
      request.path = (request.path || []).concat(pathParameter);
    } else if (isHeaderParameter(parameter)) {
      const headerParameter = translateToHeaderParam(parameter, consumes);
      request.headers = (request.headers || []).concat(headerParameter);
    }
    return request;
  }, {});
}
