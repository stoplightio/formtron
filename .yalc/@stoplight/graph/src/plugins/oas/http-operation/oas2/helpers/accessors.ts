import get = require('lodash/get');
import pick = require('lodash/pick');
import values = require('lodash/values');
import { Operation, Security, Spec } from 'swagger-schema-official';
import { uniqFlatMap } from '../../common/utils';

export function getSecurities(operation: Partial<Operation>, spec: Partial<Spec>): Security[] {
  const globalSecuritySchemes = spec.security || [];
  const schemes = uniqFlatMap(globalSecuritySchemes);
  const globalSecurity = values(pick(spec.securityDefinitions || {}, schemes));
  return operation.security || globalSecurity;
}

function getProducesOrConsumes(which: 'produces' | 'consumes', operation: Partial<Operation>, spec: Partial<Spec>) {
  const mimeTypes = get(operation, which, get(spec, which, [])) as string[];
  return mimeTypes.length ? mimeTypes : ['*'];
}

export function getProduces(operation: Partial<Operation>, spec: Partial<Spec>) {
  return getProducesOrConsumes('produces', operation, spec);
}

export function getConsumes(operation: Partial<Operation>, spec: Partial<Spec>) {
  return getProducesOrConsumes('consumes', operation, spec);
}
