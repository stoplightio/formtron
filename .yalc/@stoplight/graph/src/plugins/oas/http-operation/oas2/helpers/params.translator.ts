import {
  HttpParamStyles,
  IHttpEncoding,
  IHttpHeaderParam,
  IHttpPathParam,
  IHttpQueryParam,
  IHttpRequestBody,
} from '@stoplight/types';
import partial = require('lodash/partial');
import {
  BodyParameter,
  FormDataParameter,
  Header,
  HeaderParameter,
  Parameter,
  PathParameter,
  QueryParameter,
} from 'swagger-schema-official';
import { mapDict } from '../../common/utils';
import { translateToContent } from './content.translator';

function chooseQueryParameterStyle(
  parameter: QueryParameter
): HttpParamStyles.PipeDelimited | HttpParamStyles.SpaceDelimited | HttpParamStyles.Form {
  /** Must cast to 'any' because this field is missing from the types but it's defined in the spec
   * https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#parameterObject
   */
  switch ((parameter as any).collectionFormat) {
    case 'pipes':
      return HttpParamStyles.PipeDelimited;
    case 'ssv':
      return HttpParamStyles.SpaceDelimited;
    case 'csv':
    case 'multi':
    default:
      /**
       * This implementation is, in fact, not fully compliant with oas3.
       * As per oas3 spec: "Form style parameters defined by RFC6570.
       *                    This option replaces collectionFormat with a csv
       *                    (when explode is false) or multi (when explode is true)
       *                    value from OpenAPI 2.0."
       * But since there is no such property like 'explode' in oas2 we are defaulting to 'form'.
       */
      return HttpParamStyles.Form;
  }
}

function commonFields(parameter: Parameter) {
  return {
    description: parameter.description,
    required: parameter.required,
  };
}

function translatePartiallyToHeaderParam(mediaTypes: string[], header: Header, name: string): IHttpHeaderParam {
  return {
    description: header.description,
    // `required` not suppored by oas2
    name,
    content: translateToContent(header, mediaTypes),
    style: HttpParamStyles.Simple,
    // `deprecated` not supported by oas2
    // `explode` not supported by oas2,
  };
}

export function translateToHeaderParam(parameter: HeaderParameter, consumes: string[]): IHttpHeaderParam {
  return {
    required: parameter.required,
    ...translatePartiallyToHeaderParam(consumes, parameter, parameter.name),
  };
}

export function translateToHeaderParams(
  headers: { [headerName: string]: Header },
  consumes: string[]
): IHttpHeaderParam[] {
  return mapDict(headers, partial(translatePartiallyToHeaderParam, consumes));
}

export function translateToBodyParameter(parameter: BodyParameter, consumes: string[]): IHttpRequestBody {
  return {
    ...commonFields(parameter),
    contents: consumes.map(mediaType => ({
      mediaType,
      schema: parameter.schema,
      // 'encoding' not supported by oas2
      // 'examples' not supported by oas2
    })),
  };
}

export function translateToFormDataParameter(
  parameter: FormDataParameter,
  requestBody: IHttpRequestBody | null | undefined,
  consumes: string[]
): IHttpRequestBody {
  const requestBodyCopy: IHttpRequestBody = Object.assign(
    {},
    requestBody || {
      ...commonFields(parameter),
      contents: consumes.map(mediaType => ({
        mediaType,
        schema: parameter,
        // 'encoding' not supported by oas2
        // 'examples' not supported by oas2
      })),
    }
  );
  const bodyContent = requestBodyCopy.contents[0];
  const encoding: IHttpEncoding = {
    property: parameter.name,
    style: HttpParamStyles.Form,
    // `mediaType` not supported by oas2
    // `allowReserved` not supported by oas2
    // `explode` not supported by oas2
    // `mediaType` not supported by oas2
    // `headers` not supported by oas2
  };
  bodyContent.encodings = (bodyContent.encodings || []).concat(encoding);
  return requestBodyCopy;
}

export function translateToQueryParameter(parameter: QueryParameter, consumes: string[]): IHttpQueryParam {
  return {
    ...commonFields(parameter),
    allowEmptyValue: parameter.allowEmptyValue,
    name: parameter.name,
    content: translateToContent(parameter, consumes),
    style: chooseQueryParameterStyle(parameter),
    // `allowReserved` not supported by oas2
    // `deprecated` not supported by oas2
    // `explode` not supported by oas2,
  };
}

export function translateToPathParameter(parameter: PathParameter, consumes: string[]): IHttpPathParam {
  return {
    ...commonFields(parameter),
    name: parameter.name,
    content: translateToContent(parameter, consumes),
    // TODO(SL-248): as far as I can see in the oas3 spec simple is the only valid style, all other styles
    // defined in oas2 are obsolete (doubt: what about 'pipes', 'tsv' and 'ssv')
    style: HttpParamStyles.Simple,
    // `deprecated` not supported by oas2
    // `explode` not supported by oas2,
  };
}
