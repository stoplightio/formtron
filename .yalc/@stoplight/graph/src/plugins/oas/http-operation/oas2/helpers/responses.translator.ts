import { IHttpResponse } from '@stoplight/types';
import partial = require('lodash/partial');
import { Response } from 'swagger-schema-official';
import { mapDict } from '../../common/utils';
import { translateToHeaderParams } from './params.translator';

const toObject = <T>(value: T, key: string) => ({ key, value });

function translateToResponse(produces: string[], response: Response, statusCode: string) {
  const headers = translateToHeaderParams(response.headers || {}, produces);
  return {
    code: statusCode,
    description: response.description,
    headers,
    contents: produces.map(mediaType => ({
      mediaType,
      schema: response.schema,
      examples: mapDict(response.examples, toObject),
      // `encoding` is not supported
    })),
    // `links` not supported by oas2
  } as IHttpResponse;
}

export function translateToResponses(responses: { [name: string]: Response }, produces: string[]): IHttpResponse[] {
  return mapDict(responses, partial(translateToResponse, produces));
}
