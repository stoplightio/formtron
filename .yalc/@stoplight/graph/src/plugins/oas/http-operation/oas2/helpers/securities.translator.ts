import {
  IApiKeySecurityScheme,
  IBasicSecurityScheme,
  IOauth2SecurityScheme,
  IOauthFlowObjects,
} from '@stoplight/types';
import { ApiKeySecurity, BaseOAuthSecuirty, Security } from 'swagger-schema-official';
import { HttpSecurityScheme } from '../../common/types';

import compact = require('lodash/compact');

/**
 * @param security the union with 'any' is purposeful. Passing strict types does not help much here,
 * because all these checks happen in runtime. I'm leaving 'Security' only for visibility.
 */
function translateToFlows(security: Security | any): IOauthFlowObjects {
  const tokenAndScope = {
    tokenUrl: security.tokenUrl,
    scopes: security.scopes,
  };

  const flowsDict = {
    implicit: ['implicit', { scopes: [], authorizationUrl: security.authorizationUrl }],
    password: ['password', tokenAndScope],
    application: ['clientCredentials', tokenAndScope],
    accessCode: ['authorizationCode', { ...tokenAndScope, authorizationUrl: security.authorizationUrl }],
  };

  const flow = flowsDict[security.flow];
  return flow ? { [flow[0]]: flow[1] } : {};
}

function translateToBasicSecurityScheme(security: Security): IBasicSecurityScheme {
  return {
    type: 'http',
    scheme: 'basic',
    description: security.description,
  };
}

function translateToApiKeySecurityScheme(security: ApiKeySecurity): IApiKeySecurityScheme {
  const acceptableSecurityOrigins = ['query', 'header', 'cookie'];
  if (!acceptableSecurityOrigins.includes(security.in)) {
    throw new Error(
      `Provided security origin (the 'in' property): '${
        security.in
      }' is not valid. Should be one of the following: ${acceptableSecurityOrigins}`
    );
  }
  return {
    type: 'apiKey',
    name: security.name,
    in: security.in as 'query' | 'header' | 'cookie',
    description: security.description,
  };
}

function translateToOauth2SecurityScheme(security: BaseOAuthSecuirty): IOauth2SecurityScheme {
  return {
    type: 'oauth2',
    flows: translateToFlows(security),
    description: security.description,
  };
}

function translateToSingleSecurity(security: Security) {
  switch (security.type) {
    case 'basic':
      return translateToBasicSecurityScheme(security);
    case 'apiKey':
      return translateToApiKeySecurityScheme(security as ApiKeySecurity);
    case 'oauth2':
      return translateToOauth2SecurityScheme(security as BaseOAuthSecuirty);
  }
  return null;
}

export function translateToSecurities(securities: Security[]): HttpSecurityScheme[] {
  return compact(securities.map(translateToSingleSecurity));
}
