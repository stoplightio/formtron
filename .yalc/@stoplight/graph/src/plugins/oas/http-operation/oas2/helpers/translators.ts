import { IHttpOperation, IServer } from '@stoplight/types';
import isNull = require('lodash/isNull');
import omitBy = require('lodash/omitBy');
import { Operation, Parameter, Response, Spec } from 'swagger-schema-official';
import URI = require('urijs');
import { getOasParameters } from '../../common/accessors';
import { translateToTags } from '../../common/tag.translator';
import { Oas2TranslateOperationOptions } from '../../common/types';
import { getConsumes, getProduces, getSecurities } from './accessors';
import { translateToRequest } from './request.translator';
import { translateToResponses } from './responses.translator';
import { translateToSecurities } from './securities.translator';

export function translateToServers(operation: Partial<Operation>, spec: Partial<Spec>): IServer[] {
  const schemes = operation.schemes || spec.schemes || [];
  const { host, basePath } = spec;
  if (!host || !basePath) {
    return [];
  }
  return schemes.map(scheme => {
    return {
      url: URI()
        .scheme(scheme)
        .host(host)
        .path(basePath)
        .toString(),
    };
  });
}

export function translateOperation(opts: Oas2TranslateOperationOptions): IHttpOperation {
  const { operation, method, path, pathObject, spec } = opts;
  const produces = getProduces(operation, spec);
  const consumes = getConsumes(operation, spec);
  const httpOperation: IHttpOperation = {
    // TODO(SL-248): what shall we do with id?
    id: '?http-operation-id?',
    iid: operation.operationId,
    description: operation.description,
    deprecated: operation.deprecated,
    method,
    path,
    summary: operation.summary,
    responses: translateToResponses(operation.responses as { [name: string]: Response }, produces),
    servers: translateToServers(operation, spec),
    request: translateToRequest(
      getOasParameters(operation.parameters as Parameter[], pathObject.parameters as Parameter[]),
      consumes
    ),
    tags: translateToTags(operation.tags || []),
    security: translateToSecurities(getSecurities(operation, spec)),
  };

  return omitBy(httpOperation, isNull) as IHttpOperation;
}
