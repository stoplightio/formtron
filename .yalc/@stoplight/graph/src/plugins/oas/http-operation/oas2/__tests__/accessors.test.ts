import { getConsumes, getProduces, getSecurities } from '../helpers/accessors';

describe('accessors', () => {
  describe('getSecurities', () => {
    test('should return empty array', () => {
      expect(getSecurities({}, {})).toMatchSnapshot();
    });

    test('should fallback to operation security', () => {
      expect(getSecurities({ security: [{ type: 'basic' }] }, {})).toMatchSnapshot();
    });

    test('should fallback to global security', () => {
      expect(
        getSecurities(
          {},
          {
            security: [{ s1: [], s2: [] }, { s1: [], s3: [] }],
            securityDefinitions: {
              s1: { type: 'basic' },
              s3: { type: 'oauth2' },
              s5: { type: 'apiKey' },
            },
          }
        )
      ).toMatchSnapshot();
    });

    test('should prefer to operation security', () => {
      expect(
        getSecurities(
          { security: [{ type: 'apiKey' }] },
          {
            security: [{ s1: [] }],
            securityDefinitions: { s1: { type: 'basic' } },
          }
        )
      ).toMatchSnapshot();
    });
  });

  describe('getProduces', () => {
    test('given all empty arrays should return asterisk', () => {
      expect(getProduces({}, {})).toMatchSnapshot();
      expect(getProduces({ produces: [] }, { produces: [] })).toMatchSnapshot();
    });

    test('should fallback to spec produces', () => {
      expect(getProduces({}, { produces: ['text/plain'] })).toMatchSnapshot();
    });

    test('should prefer operation produces', () => {
      expect(getProduces({ produces: ['text/poem'] }, { produces: ['text/plain'] })).toMatchSnapshot();
    });
  });

  describe('getConsumes', () => {
    test('given all empty arrays should return asterisk', () => {
      expect(getConsumes({}, {})).toMatchSnapshot();
      expect(getConsumes({ consumes: [] }, { consumes: [] })).toMatchSnapshot();
    });

    test('should fallback to spec consumes', () => {
      expect(getConsumes({}, { consumes: ['text/plain'] })).toMatchSnapshot();
    });

    test('should prefer operation consumes', () => {
      expect(getConsumes({ consumes: ['text/poem'] }, { consumes: ['text/plain'] })).toMatchSnapshot();
    });
  });
});
