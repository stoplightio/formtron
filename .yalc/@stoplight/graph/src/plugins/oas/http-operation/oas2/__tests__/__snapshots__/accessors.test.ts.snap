// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessors getConsumes given all empty arrays should return asterisk 1`] = `
Array [
  "*",
]
`;

exports[`accessors getConsumes given all empty arrays should return asterisk 2`] = `
Array [
  "*",
]
`;

exports[`accessors getConsumes should fallback to spec consumes 1`] = `
Array [
  "text/plain",
]
`;

exports[`accessors getConsumes should prefer operation consumes 1`] = `
Array [
  "text/poem",
]
`;

exports[`accessors getProduces given all empty arrays should return asterisk 1`] = `
Array [
  "*",
]
`;

exports[`accessors getProduces given all empty arrays should return asterisk 2`] = `
Array [
  "*",
]
`;

exports[`accessors getProduces should fallback to spec produces 1`] = `
Array [
  "text/plain",
]
`;

exports[`accessors getProduces should prefer operation produces 1`] = `
Array [
  "text/poem",
]
`;

exports[`accessors getSecurities should fallback to global security 1`] = `
Array [
  Object {
    "type": "basic",
  },
  Object {
    "type": "oauth2",
  },
]
`;

exports[`accessors getSecurities should fallback to operation security 1`] = `
Array [
  Object {
    "type": "basic",
  },
]
`;

exports[`accessors getSecurities should prefer to operation security 1`] = `
Array [
  Object {
    "type": "apiKey",
  },
]
`;

exports[`accessors getSecurities should return empty array 1`] = `Array []`;
